[clang-tidy] add modernize-use-std-numbers

5chmidti opened the pull request:
Finds constants and function calls to math functions that can be replaced
with c++20's mathematical constants from the 'numbers' header and
offers fix-it hints.
Does not match the use of variables with that value, and instead,
offers a replacement at the definition of those variables.

5chmidti committed:
[clang-tidy] add modernize-use-std-numbers check

This check finds constants and function calls to math functions that can be replaced
with c++20's mathematical constants ('numbers' header) and offers fixit-hints.
Does not match the use of variables or macros with that value and instead, offers a replacement
at the definition of said variables and macros.

5chmidti commented:
For macros I have only implemented matching literals, not function calls or formulas. I also don't suggest using modern language constructs for macros, there are other checks that do this.

PiotrZSL commented:
In general scope of the check is fine, but specially as it's limited to C++20 and way how its written would be good to rename it into modernize-use-std-numbers.

5chmidti commented:
- How do I correctly parse the values in a macro? While the test locally works, it doesn't in CI. I also don't differentiate between what kind of numeric constant the token is, which might be another problem. Any pointers on how to do this correctly are greatly appreciated.
- Macros currently don't distinguish between `float`, `double` and `long double` like the AST-based check does. The values matched in the PPCallback should be able to provide the `..._v<>` versions of replacements as well. The question about how to detect the difference is probably related to the point above.

5chmidti commented:
> In general scope of the check is fine, but specially as it's limited to C++20 and way how its written would be good to rename it into modernize-use-std-numbers.

Sure thing. I don't have a strong feeling about the category.

PiotrZSL commented:
> Sure thing. I don't have a strong feeling about the category.

I'ts not only about category but also name, check suggests std::numbers, then better name it modernize-use-std-numbers, reason for that is simple. Before C++20 we may want to have a check for example <something>-prefer-math-constants that would change 3.14 hardcoded in code into M_PI. And in same way we could have check <something>-use-std-numeric-limits that would change -2147483648 into std::numeric_limits<std::sint32_t>::min(). Simply we may have many of such checks.

PiotrZSL commented:
Simply better to reduce scope of check name to avoid overlap, unless check would support pre-c++20, then it could be fine.

5chmidti commented:
> I'ts not only about category but also name, check suggests std::numbers, then better name it modernize-use-std-numbers, reason for that is simple. Before C++20 we may want to have a check for example -prefer-math-constants that would change 3.14 hardcoded in code into M_PI. And in same way we could have check -use-std-numeric-limits that would change -2147483648 into std::numeric_limitsstd::sint32_t::min(). Simply we may have many of such checks.

Sorry, I read too fast it seems. But no objections to the name change on my end. It is more precise.
I'll force push one more time before any review comments come in. 

I don't plan to implement pre c++20 macros or limits here.
However, I do plan to write a check for numeric limits, as it is quite similar to this one.

EugeneZelenko commented on clang-tools-extra/clang-tidy/modernize/UseStdNumbersCheck.cpp:
```diff
@@ -0,0 +1,377 @@
+//===--- UseStdNumbersCheck.cpp - clang_tidy -------------------------===//
```
Please make it same length as similar line at bottom of comment.

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -168,6 +168,12 @@ New checks
+  <clang-tidy/checks/modernize/use-std-numbers>` check.
+
+  Finds constants and function calls to math functions that can be replaced
+  with c++20's mathematical constants ('numbers' header).
```
Please use double back-ticks for `numbers`.

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-numbers.rst:
```diff
@@ -0,0 +1,25 @@
+modernize-use-std-numbers
+=========================
+
+This check finds constants and function calls to math functions that can be replaced
```
Please synchronize first statement with Release Notes.

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-numbers.rst:
```diff
@@ -0,0 +1,25 @@
+=========================
+
+This check finds constants and function calls to math functions that can be replaced
+with c++20's mathematical constants ('numbers' header) and offers fixit-hints.
```
`fix-it`.

5chmidti committed:
address comments

- changed ReleaseNotes and first sentence of the check documentation
  by removing parentheses around the numbers header and noted that
  fix-it hints are emitted

5chmidti committed:
update descriptions slightly

- update check description in header to reflect release notes description
- add comma

5chmidti committed:
fixup update of warning message

5chmidti committed:
rename isMathMacro to isMacro

5chmidti committed:
add tests for uses in templates

- dont match in template instantiations

5chmidti committed:
fix matcher names

5chmidti commented:
> * How do I correctly parse the values in a macro? While the test locally works, it doesn't in CI. I also don't differentiate between what kind of numeric constant the token is, which might be another problem. Any pointers on how to do this correctly are greatly appreciated.
> * Macros currently don't distinguish between `float`, `double` and `long double` like the AST-based check does. The values matched in the PPCallback should be able to provide the `..._v<>` versions of replacements as well. The question about how to detect the difference is probably related to the point above.

Ping.

5chmidti committed:
fix name of custom matcher hasCanonicalTypeUnqualified

5chmidti committed:
remove custom constants in favor of using llvm::numbers for all

5chmidti committed:
remove not needed argument for using a compile-time constant

5chmidti committed:
fix inv_pi and phi matchers matching uses of named constants

PiotrZSL commented:
> * How do I correctly parse the values in a macro? While the test locally works, it doesn't in CI. I also don't differentiate between what kind of numeric constant the token is, which might be another problem. Any pointers on how to do this correctly are greatly appreciated.
> * Macros currently don't distinguish between `float`, `double` and `long double` like the AST-based check does. The values matched in the PPCallback should be able to provide the `..._v<>` versions of replacements as well. The question about how to detect the difference is probably related to the point above.

I think you can't, instead of handling macro definitions, you just should handle places where macro is used.
It's fine if someone define a macro MY_PI, check should flag usage of such macro, not necessarily a macro.

Instead of using MacroDefined and try to find those #define MY_PI, you should just use MacroExpands, and find usage of those system macros like M_PI, and suggest replacement usage of them without checking an value.

As for #define MY_PI 3.14, this should be handled like it is now, simply isMacro shouldn't be used, and warning should be produced inside macro.

5chmidti commented:
> I think you can't, instead of handling macro definitions, you just should handle places where macro is used. It's fine if someone define a macro MY_PI, check should flag usage of such macro, not necessarily a macro.
> 
> Instead of using MacroDefined and try to find those #define MY_PI, you should just use MacroExpands, and find usage of those system macros like M_PI, and suggest replacement usage of them without checking an value.
> 
> As for #define MY_PI 3.14, this should be handled like it is now, simply isMacro shouldn't be used, and warning should be produced inside macro.

Sounds good to do it like this for macros. My original thought about this was to be the as least intrusive as possible in terms of the number of matches and fixes this check does, but I'm all for replacing these macros.
Should variables be treated the same way? I.e. should the fix for
```c++
static constexpr double Phi = 1.6180339;
sink(Phi);
```
be changed from
```c++
static constexpr double Phi = std::numbers::phi;
sink(Phi)
```
to 
```c++
static constexpr double Phi = std::numbers::phi;
sink(std::numbers::phi)
```
?

I could add an option `MatchVariableUses` for this:
- Never (the current way)
- (LocalVariables (matches only uses of variables that are (function-)local declared))?
- Always (like the above proposed last change of the fixits)

defaulting to...

PiotrZSL commented:
No need to change to `sink(std::numbers::phi)`, simply just replace numbers, leave usage like it is.

5chmidti committed:
change matches of macro defs to macro uses

- modernizes the usage of macros to the std::numbers constants
- removes PPCallback that failed CI
- change comments and docs

5chmidti committed:
split matcher for float values

- removes the optional bool flag
- improves readability for matchers
- split the comment

5chmidti committed:
add test of using math marco on its own

5chmidti committed:
add newline after code-block directive to maybe fix docs ci

5chmidti committed:
fix APFloat issue

5chmidti committed:
run clang-include-cleaner

5chmidti committed:
rename chainedIfBound to applyRuleForBoundOrDefault

5chmidti commented:
I rebased and force-pushed my branch to fix the formatting CI (fetch depth for comparison with head was too big).
I pushed no actual changes in that force-push (one documentation ordering change in `fix lexicographical ordering in some places after rename` became redundant).
@EugeneZelenko's comments were already addressed and, even without the force-push, had already lost the context.

I'm not sure what the issue on Windows is. It looks like nothing within the function template `baz` is matching.

5chmidti commented:
Nice, the rebase fixed the Windows failure

5chmidti committed:
add support for combinations of casts

PiotrZSL commented on clang-tools-extra/clang-tidy/modernize/UseStdNumbersCheck.cpp:
```diff
@@ -0,0 +1,284 @@
+
+AST_MATCHER_P(clang::QualType, hasCanonicalTypeUnqualified,
+              Matcher<clang::QualType>, InnerMatcher) {
+  return InnerMatcher.matches(Node->getCanonicalTypeUnqualified(), Finder,
```
Verify here that Node is Null, so `return !Node.isNull() && InnerMatcher...`

5chmidti replied:
done

PiotrZSL commented on clang-tools-extra/clang-tidy/modernize/UseStdNumbersCheck.cpp:
```diff
@@ -0,0 +1,284 @@
+                              Builder);
+}
+
+AST_MATCHER(clang::QualType, isArithmetic) { return Node->isArithmeticType(); }
```
Same in those both, verify that Type isn't null.

5chmidti replied:
done

PiotrZSL commented on clang-tools-extra/clang-tidy/modernize/UseStdNumbersCheck.cpp:
```diff
@@ -0,0 +1,284 @@
+
+  const auto Dref = declRefExpr(to(varDecl(
+      anyOf(isConstexpr(),
+            varDecl(hasType(qualType(isConstQualified(), isArithmetic())))),
```
Consier checking here CanonicalType to ignore all typedefs

5chmidti replied:
This is not needed. I just checked https://github.com/llvm/llvm-project/blob/af4abc4fa7778f83ba7fd759b818352dc2d19469/clang/lib/AST/Type.cpp#L2186 The canonical type is used to check if the type is a floating type, instead of the `BaseType` pointer in `ExtQualsTypeCommonBase`.

Tests that confirm this behavior are
- `bar::sqrt(Actually2MyFloat);`
- `log2(Euler5);` (`Euler5` has the type `my_double`)
- `log2(Euler6);` (`Euler6` has the type `my_float`)

PiotrZSL commented on clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-numbers.rst:
```diff
@@ -0,0 +1,28 @@
+with c++20's mathematical constants from the ``numbers`` header and offers fix-it hints.
+Does not match the use of variables with that value, and instead,
+offers a replacement at the definition of those variables.
+
```
List here all possible replacements.

5chmidti replied:
done

5chmidti replied:
I’ve also extended the documentation slightly to explain the replacements a bit further.

5chmidti committed:
check QualType::isNull in custom matchers

5chmidti committed:
refactor Dref match in matchFloatValueNear

- constexpr implies const
- constrain the varDecl's type to be a floating type
- use ignoreImplicitAndFloatingCasting on the hasInitializer expression

5chmidti committed:
change Dref match in MatchValue

- constexpr implies const

5chmidti committed:
fix implicit casts in hasInitializer of matchValue

- implicit conversiosn might occur
- allow only implicit casts to floating types
when the initializer is a float literal
- allow implicit conversions for ints

5chmidti committed:
add test for int constants initialized with float

5chmidti committed:
extend documentation

- provide the list of supported constants
- explain replacements a bit more

PiotrZSL reviewed (changes requested):
Example:
```
llvm/include/llvm/Support/MathExtras.h:59:31: warning: prefer std::numbers math constant [modernize-use-std-numbers]
   59 |                 inv_sqrt3f  = .577350269F, // (0x1.279a74P-1)
      |                               ^~~~~~~~~~~
      |                               std::numbers::egamma_v<float>
```

```
egammaf     = .577215665F
```

Looks like having this check implemented as an multiple matchers isn't a good idea, simply because we pickup first one that match instead a nearest one. This leads to bugs when dealing with proper values.

In ideal conditions something like x* 3.14 should be even detected as PI. 
Also warning message should already say what from std::numbers should be used and how far are current and proposed values from them self.

also this precision 0.001 should be put into configuration option.
On bright side, works fine for a M_PI macros and properly add missing header.

5chmidti committed:
refactor from using transformer to a normal check

- change from seperated matchers to a single unified matcher
- prefer to diagnose patterns (formulas) before literals,
  diagnosing the expr with the largest possible impact/fixit
- the numbers constant that is the closest match to the matched literal
  is offered as a replacement, instead of the first one found.
- explicitly mention the constant that the fix-it is offering

5chmidti committed:
change message to include what is being replaced and how far off literals are

5chmidti committed:
add config value for DiffThreshold

5chmidti committed:
use TK_IgnoreUnlessSpelledInSource

- makes some matchers a little bit more readable
- removes the need for the double `ignoringImplicit`
  in the custom ignoringParenAnd_Casting

5chmidti committed:
rm duplicate canonicalization of expr type

5chmidti committed:
rm not needed parens

5chmidti committed:
declare PatternBindString in the if-init

5chmidti committed:
more const parameters

5chmidti committed:
init DiffThresholdString in the init list

5chmidti committed:
add comment for isRangeOfCompleteMacro

5chmidti committed:
add explicit tests for macro range check

5chmidti committed:
instantiate big test template

5chmidti commented:
@PiotrZSL I have implemented the changes you have requested, but because of #72536, I am not going to push those changes yet, unless you prefer it if I do. I might take a look into fixing this issue.
I came across this when I checked what would happen when the big templated function was instantiated, just to be sure the test would pass.

PiotrZSL commented:
> @PiotrZSL I have implemented the changes you have requested, but because of #72536, I am not going to push those changes yet, unless you prefer it if I do. I might take a look into fixing this issue. I came across this when I checked what would happen when the big templated function was instantiated, just to be sure the test would pass.

TK_IgnoreUnlessSpelledInSource should do a trick. I will re-check this review in some near free time.

5chmidti committed:
fix comment

5chmidti committed:
rm unless(isInTemplateInstantiation()), because of IgnoreUnlessSpelledInSource

5chmidti committed:
fix literal in template problem in exchange for not resolving implicit casts

5chmidti committed:
move def of storeOptions to cpp file

5chmidti committed:
remove unused include

5chmidti commented:
It looks like GitHub decided to not list all new commits in the conversation view... there are a few more in the commits tab.
The newly pushed changes start at commit [refactor from using transformer to a normal check](https://github.com/llvm/llvm-project/pull/66583/commits/b1adc10d5a0b1456725979a1f480a517b37b6b2f).

The reason I found the issue I have created is what I had to change/remove in this commit:
[fix literal in template problem in exchange for not resolving implicit casts](https://github.com/llvm/llvm-project/pull/66583/commits/01d6b2e4e01e06e894710756ee83c668995bafbb)
I wanted the replacements to resolve implicit casts, like they do for explicit casts currently. The test shows the (IMO) degraded quality of the replacements.
After my findings/comment in that issue, I decided to push my changes this way to maybe move forward with this check and leave the implicit cast matching for another pr in the future. What do you think?

> Looks like having this check implemented as an multiple matchers isn't a good idea, simply because we pickup first one that match instead a nearest one. This leads to bugs when dealing with proper values.

There is a single matcher that matches literals and formulas. Formulas are applied as soon as a matched pattern is found. All matched literals are collected and sorted, if no formula was matched and the literal that differs the least from its constant is selected. 

> In ideal conditions something like x* 3.14 should be even detected as PI.

and

> also this precision 0.001 should be put into configuration option.

This is now configurable with the config value `DiffThreshold`. The default value is `0.001`, like it was before. Should this default be lower to detect your `3.14` example?

> Also warning message should already say what from std::numbers should be used and how far are current and proposed values from them self.

The messages now follow the following format:

- `prefer 'std::numbers::pi' to this literal, differs by '2.34e-2'` (using scientific notation to be more compact than printing `0.00000000123456789`)
- `prefer 'std::numbers::pi_v<float>' to this formula` (the message uses the replacement code from the fix-it hint)
- `prefer 'std::numbers::pi' to this macro, differs by '2.34e-2'` (if we are replacing a macro that is just a literal)
- `prefer 'std::numbers::inv_sqrt3' to this macro` (for things like `#define INV_SQRT3 1 / bar::sqrt(3)`)

So the messages include what kind of symbol/code is being replaced, the inserted code, and the difference to the inserted value if the replaced code is a literal or a macro that is a literal. The difference that is being printed is the absolute difference. Does the sign matter?

5chmidti commented on clang-tools-extra/test/clang-tidy/checkers/modernize/use-std-numbers.cpp:
```diff
@@ -128,169 +128,169 @@ void foo(){
+    // CHECK-MESSAGES: :[[@LINE-1]]:10: warning: prefer 'std::numbers::inv_sqrt3' math constant [modernize-use-std-numbers]
     // CHECK-FIXES: sink(std::numbers::inv_sqrt3);
 
+    const auto inv_sqrt3f = .577350269F;
```
> Example: [...]

I have added a test for this case

PiotrZSL reviewed (approved):
LGTM, this should be merged so other llvm users could test it.
As check is complicated I expect that there can be some false-positives or issues reported when llvm 18 would release.

5chmidti committed:
add options to doc

5chmidti commented:
Force push to resolve conflicts

5chmidti commented:
I fetched from the wrong remote. Conflicts are resolved. The options docs were added in 284fa1b94ef72c29f13a2ec50ca76b429c842b0c

PiotrZSL commented on clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-numbers.rst:
```diff
@@ -0,0 +1,74 @@
+
+    A floating point value that sets the detection threshold for when literals
+    match a constant.
+    A literal matches a constant if
```
Continue in previous line.

5chmidti replied:
meant to write done, not click resolve

PiotrZSL commented on clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-numbers.rst:
```diff
@@ -0,0 +1,74 @@
+    A floating point value that sets the detection threshold for when literals
+    match a constant.
+    A literal matches a constant if
+    ``abs(literal - constant) < DiffThreshold`` evaluates to true.
```
true -> `true`

PiotrZSL reviewed (approved):
Unless you want this to be merged as 44101708+5chmidti@users.noreply.github.com, correct your github settings, squash all commits into one and make sure that author of commit is correct.

5chmidti committed:
address doc comments

5chmidti commented:
That email is fine with me

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -193,6 +193,13 @@ New checks
+  <clang-tidy/checks/modernize/use-std-numbers>` check.
+
+  Finds constants and function calls to math functions that can be replaced
+  with c++20's mathematical constants from the ``numbers`` header and
```
`C++20`.

5chmidti replied:
Do you just want the `C` to be big, or do you want the version inside of single or double `?

EugeneZelenko replied:
Uppercase `C`. See https://en.cppreference.com.

5chmidti replied:
Thx, done

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-numbers.rst:
```diff
@@ -0,0 +1,73 @@
+=========================
+
+Finds constants and function calls to math functions that can be replaced
+with c++20's mathematical constants from the ``numbers`` header and offers
```
Ditto.

5chmidti replied:
done

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-numbers.rst:
```diff
@@ -0,0 +1,73 @@
+
+The following list of constants from the ``numbers`` header are supported:
+
+* e
```
Please enclose each constant in double back-ticks.

5chmidti replied:
done

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-numbers.rst:
```diff
@@ -0,0 +1,73 @@
+The list currently includes all constants as of C++20.
+
+The replacements use the type of the matched constant and can remove explicit casts,
+i.e., switching between ``std::numbers::e``, ``std::numbers::e_v<float>`` and ``std::numbers::e_v<long double>``
```
Please follow 8- characters limit.

5chmidti committed:
C++20

5chmidti committed:
double backtick elements in constants list

5chmidti committed:
reflow docs for 80 char column limit

5chmidti committed:
change at to for in sentence

5chmidti commented on clang-tools-extra/docs/clang-tidy/checks/modernize/use-std-numbers.rst:
```diff
@@ -5,7 +5,7 @@ modernize-use-std-numbers
 Finds constants and function calls to math functions that can be replaced
 with c++20's mathematical constants from the ``numbers`` header and offers fix-it hints.
-Does not match the use of variables or macros with that value and instead, offers a replacement
+Does not match the use of variables or macros with that value, and instead, offers a replacement
```
This looks like an older review comment from your previous review was probably saved by GitHub. The comma exists in the current version and the column limit is not exceeded.

5chmidti commented:
Thank you for the reviews

5chmidti commented:
@PiotrZSL check out https://lab.llvm.org/buildbot/#/builders/230/builds/22226, the bot was red on previous runs, but those are unrelated to how it fails now. The bot fails for this patch with the following build failure:
```
/home/buildbots/ppc64le-clang-test-suite/clang-ppc64le-test-suite/llvm-project/clang-tools-extra/clang-tidy/modernize/UseStdNumbersCheck.cpp: In member function ‘virtual void clang::tidy::modernize::UseStdNumbersCheck::registerMatchers(clang::ast_matchers::MatchFinder*)’:
/home/buildbots/ppc64le-clang-test-suite/clang-ppc64le-test-suite/llvm-project/clang-tools-extra/clang-tidy/modernize/UseStdNumbersCheck.cpp:319:27: error: invalid initialization of reference of type ‘const llvm::ArrayRef<clang::ast_matchers::internal::Matcher<clang::Stmt> >&’ from expression of type ‘const std::initializer_list<const clang::ast_matchers::internal::Matcher<clang::Stmt> >’
           anyOfExhaustive(ConstantMatchers),
                           ^~~~~~~~~~~~~~~~
In file included from /home/buildbots/ppc64le-clang-test-suite/clang-ppc64le-test-suite/llvm-project/clang/include/clang/ASTMatchers/ASTMatchers.h:73,
                 from /home/buildbots/ppc64le-clang-test-suite/clang-ppc64le-test-suite/llvm-project/clang/include/clang/ASTMatchers/ASTMatchFinder.h:43,
                 from /home/buildbots/ppc64le-clang-test-suite/clang-ppc64le-test-suite/llvm-project/clang-tools-extra/clang-tidy/modernize/../ClangTidyCheck.h:14,
                 from /home/buildbots/ppc64le-clang-test-suite/clang-ppc64le-test-suite/llvm-project/clang-tools-extra/clang-tidy/modernize/UseStdNumbersCheck.h:12,
                 from /home/buildbots/ppc64le-clang-test-suite/clang-ppc64le-test-suite/llvm-project/clang-tools-extra/clang-tidy/modernize/UseStdNumbersCheck.cpp:9:
/home/buildbots/ppc64le-clang-test-suite/clang-ppc64le-test-suite/llvm-project/clang-tools-extra/clang-tidy/modernize/UseStdNumbersCheck.cpp:64:28: note: in passing argument 1 of ‘clang::ast_matchers::internal::Matcher<clang::Expr> {anonymous}::anyOfExhaustive(const llvm::ArrayRef<clang::ast_matchers::internal::Matcher<clang::Stmt> >&)’
 AST_MATCHER_P(clang::Expr, anyOfExhaustive,
                            ^~~~~~~~~~~~~~~
/home/buildbots/ppc64le-clang-test-suite/clang-ppc64le-test-suite/llvm-project/clang/include/clang/ASTMatchers/ASTMatchersMacros.h:149:57: note: in definition of macro ‘AST_MATCHER_P_OVERLOAD’
   inline ::clang::ast_matchers::internal::Matcher<Type> DefineMatcher(         \
                                                         ^~~~~~~~~~~~~
/home/buildbots/ppc64le-clang-test-suite/clang-ppc64le-test-suite/llvm-project/clang-tools-extra/clang-tidy/modernize/UseStdNumbersCheck.cpp:64:1: note: in expansion of macro ‘AST_MATCHER_P’
 AST_MATCHER_P(clang::Expr, anyOfExhaustive,
 ^~~~~~~~~~~~~
```

The mismatch happens for the template arguments of `ArrayRef` and `initializer_list` in `anyOfExhaustive`:
`const llvm::ArrayRef<clang::ast_matchers::internal::Matcher<clang::Stmt> >&`
`const std::initializer_list<const clang::ast_matchers::internal::Matcher<clang::Stmt> >`

I think the easiest fix is to add `const` to the `ArrayRef` in `anyOfExhaustive` like so:
```c++
AST_MATCHER_P(clang::Expr, anyOfExhaustive,
              llvm::ArrayRef<const Matcher<clang::Stmt>>, Exprs) {
```

https://github.com/llvm/llvm-project/blob/65cb5d58eb5783c17fd2121db79ebac950b767b2/clang-tools-extra/clang-tidy/modernize/UseStdNumbersCheck.cpp#L64-L75

FYI I don't have write access, so I cannot push a fix myself. Or just revert?
What do you think?

PiotrZSL commented:
I will compile this locally and try to fix it & recommit.

5chmidti commented:
Thanks
